File: sdks/ruby/test/test_connect_retry.rb
require "test/unit"
require "socket"
require "tmpdir"
require_relative "../lib/beachcomber/client"

class TestConnectRetry < Test::Unit::TestCase
  def test_connect_retries_succeed_after_brief_outage
    Dir.mktmpdir("comb-retry-") do |dir|
      sock_path = File.join(dir, "sock")

      binder = Thread.new do
        sleep 0.4
        srv = UNIXServer.new(sock_path)
        client = srv.accept
        client&.close
        srv.close
      end

      start = Time.now
      sock = Beachcomber::Client._connect_with_retry(sock_path)
      elapsed = Time.now - start

      assert_not_nil sock
      assert elapsed >= 0.25, "expected retry; elapsed=#{elapsed}"
      sock.close
      binder.join
    end
  end

  def test_connect_retries_exhaust
    Dir.mktmpdir("comb-retry-") do |dir|
      sock_path = File.join(dir, "nosock")
      start = Time.now
      assert_raise(Errno::ECONNREFUSED, Errno::ENOENT) do
        Beachcomber::Client._connect_with_retry(sock_path)
      end
      elapsed = Time.now - start
      assert elapsed >= 1.7, "expected full retry wait; elapsed=#{elapsed}"
    end
  end
end

File: sdks/python/tests/test_connect_retry.py
import socket
import threading
import time
from pathlib import Path

import pytest


def test_connect_retries_succeed_after_brief_outage(tmp_path):
    from libbeachcomber.client import _connect_with_retry

    sock_path = tmp_path / "sock"

    def binder():
        time.sleep(0.4)
        s = socket.socket(socket.AF_UNIX, socket.SOCK_STREAM)
        s.bind(str(sock_path))
        s.listen(1)
        s.settimeout(5)
        try:
            conn, _ = s.accept()
            conn.close()
        except Exception:
            pass
        s.close()

    threading.Thread(target=binder, daemon=True).start()

    start = time.time()
    sock = _connect_with_retry(str(sock_path))
    elapsed = time.time() - start

    assert sock is not None
    assert elapsed >= 0.25, f"should have retried; elapsed={elapsed}"
    sock.close()


def test_connect_retries_exhaust(tmp_path):
    from libbeachcomber.client import _connect_with_retry

    sock_path = tmp_path / "nosock"
    start = time.time()
    with pytest.raises((ConnectionRefusedError, FileNotFoundError)):
        _connect_with_retry(str(sock_path))
    elapsed = time.time() - start
    # 250 + 500 + 1000 = 1750ms minimum.
    assert elapsed >= 1.7, f"should wait through all retries; elapsed={elapsed}"
